building_map_tools/building_crowdsim/navmesh/polygon.py

```diff
@@ -0,0 +1,76 @@
+from .vertex import Vertex
+from .object_manager import Manager
+
+
+class Polygon:
+
+    def __init__(self):
+        self.id = -1
+        # store the actual vertices, vertices are ordered in sequence!
+        self.vertices = []
+        self.vertex_ids = set()
+        self.edge_ids = set()
+        self.obstacle_ids = set()
+        # gradient stands for plane equation for Ax+By+C = 0
+        self.gradient = [0, 0, 0]
+        self.center = Vertex([0, 0])
+
+    def add_vertex(self, vertex):
+        assert(isinstance(vertex, Vertex))
+        if vertex.id in self.vertex_ids:
+            return
+        self.vertex_ids.add(vertex.id)
+        self.vertices.append(vertex)
+
+    def calulate_center(self):
+        num = len(self.vertices)
+        assert(num >= 3), "Polygon has less than 3 vertices"
+        x = 0.0
+        y = 0.0
+        for vtx in self.vertices:
+            x += vtx.x
+            y += vtx.y
+        self.center = Vertex([x / num, y / num])
+
+    def get_variable(self):
+        result = []
+        self.calulate_center()
+        result.append(self.center.get_coords())
+        vertices_id = []
+        for v in self.vertices:
+            vertices_id.append(v.id)
+        result.append(vertices_id)
+        result.append(self.gradient)
+        result.append(self.edge_ids)
+        result.append(self.obstacle_ids)
+        return result
```

Comment:
I'm generally concerned with the `get_variable` functions you introduce in your classes. Imho it would be a lot clearer if downstream users just received the object and accessed the variables they need by name (i.e. `poly.gradient` or `poly.edge_ids`) instead of packing everything into a list and having to access it by index (that at the end is kind of a magic number)

Reply:
Oh, this get_variable is called when writing the actual navmesh to the file. When writing the navmesh, the manager will go over all the vertices, polygons, edges, obstacles, and write the corresponding params in it.
It's like the building.yaml. All the lanes, edges, models are construct based on the vertices. Each vertex has one unique id, and each lane only store the related vertices id rather than store the actual vertices instance. 
Similarly, polygon only store the ids of related vertices, ids of related edges, ids of related obstacles. And each edge only store the related vertices ids, the same as obstacles.
That's why the above "object_manager" is needed to assign a unique id for each added instance.